java.lang.LinkageError occured when lookup a remote EJB with multiple threads

## Environment Details
GlassFish Version (and build number): Eclipse GlassFish  6.2.1  (build master-b577-g50d68a3 2021-09-01T02:07:52+0000)
JDK version: OpenJDK 11.0.12+7
OS: Red Hat Enterprise Linux release 8.0
Database:

## Problem Description
java.lang.LinkageError occured when a remote EJB are lookuped with multiple threads.
```
Caused by: java.lang.LinkageError: loader 'app' attempted duplicate class definition for test.ejb._EJBRemote_Wrapper. (test.ejb._EJBRemote_Wrapper is in unnamed module of loader 'app')
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
        ... 19 more
```

Payara server seems to have fixed this problem as
https://github.com/payara/Payara/issues/3087

## Steps to reproduce
1. deploy EJB app
2. lookup the remote EJB with multiple threads

## Impact of Issue
Cannot lookup remote EJB with multiple threads.

## Additional info to reproduce this issue
[LinkageError.zip](https://github.com/eclipse-ee4j/glassfish/files/7167452/LinkageError.zip)

```
$ glassfish6/glassfish/bin/asadmin deploy LinkageError.jar
Application deployed with name LinkageError.
Command deploy executed successfully.
$ java -classpath glassfish6/glassfish/modules/glassfish-naming.jar:glassfish6/glassfish/lib/gf-client.jar:./LinkageError.jar test.client.SessionClient localhost 23700 5 10 0
START : MainClient
START : SessionClient
START : SessionClient
START : SessionClient
START : SessionClient
START : SessionClient
Sep 15, 2021 2:39:36 PM com.sun.enterprise.v3.server.CommonClassLoaderServiceImpl findDerbyClient
INFO: Cannot find javadb client jar file, derby jdbc driver will not be available by default.
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.glassfish.pfl.basic.reflection.Bridge (file:/glassfish6/glassfish/modules/pfl-basic.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of org.glassfish.pfl.basic.reflection.Bridge
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Exception in thread "Thread-0" java.lang.Error: Could not access ClassLoader.defineClass()
        at org.glassfish.pfl.basic.reflection.Bridge.defineClass(Bridge.java:106)
        at org.glassfish.pfl.dynamic.codegen.impl.CodeGeneratorUtil.makeClass(CodeGeneratorUtil.java:46)
        at org.glassfish.pfl.dynamic.codegen.spi.Wrapper._generate(Wrapper.java:1026)
        at org.glassfish.pfl.dynamic.codegen.spi.Wrapper._generate(Wrapper.java:1010)
        at com.sun.ejb.EJBUtils.generateAndLoad(EJBUtils.java:569)
        at com.sun.ejb.EJBUtils.loadGeneratedRemoteBusinessClasses(EJBUtils.java:483)
        at com.sun.ejb.EJBUtils.loadGeneratedRemoteBusinessClasses(EJBUtils.java:441)
        at com.sun.ejb.EJBUtils.lookupRemote30BusinessObject(EJBUtils.java:389)
        at com.sun.ejb.containers.RemoteBusinessObjectFactory.getObjectInstance(RemoteBusinessObjectFactory.java:51)
        at java.naming/javax.naming.spi.NamingManager.getObjectInstance(NamingManager.java:341)
        at com.sun.enterprise.naming.impl.SerialContext.getObjectInstance(SerialContext.java:503)
        at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:463)
        at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:414)
        at java.naming/javax.naming.InitialContext.lookup(InitialContext.java:409)
        at test.client.SessionClient.run(SessionClient.java:77)
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.glassfish.pfl.basic.reflection.Bridge.defineClass(Bridge.java:104)
        ... 14 more
Caused by: java.lang.LinkageError: loader 'app' attempted duplicate interface definition for test.ejb._EJBRemote_Remote. (test.ejb._EJBRemote_Remote is in unnamed module of loader 'app')
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
        ... 19 more
```

--- a/appserver/ejb/ejb-container/src/main/java/com/sun/ejb/EJBUtils.java
+++ b/appserver/ejb/ejb-container/src/main/java/com/sun/ejb/EJBUtils.java
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 1997, 2018 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -536,63 +537,81 @@ public class EJBUtils {
 
 
     private static Class generateAndLoad(ClassGeneratorFactory cgf,
-                                         final String actualClassName,
-                                         final ClassLoader loader,
-                                         final Class protectionDomainBase) {
-
-        cgf.evaluate();
-
-        final Properties props = new Properties();
-        if( _logger.isLoggable(Level.FINE) ) {
-
-            props.put(DUMP_AFTER_SETUP_VISITOR, "true");
-            props.put(TRACE_BYTE_CODE_GENERATION, "true");
-            props.put(USE_ASM_VERIFIER, "true");
-
-            try {
-
-                ByteArrayOutputStream baos = new ByteArrayOutputStream();
-                PrintStream ps = new PrintStream(baos);
-
-                _sourceCode(ps, props);
-                _logger.fine(baos.toString());
-
-            } catch(Exception e) {
-                _logger.log(Level.FINE, "exception generating src", e);
-            }
-
+            final String actualClassName,
+            final ClassLoader loader,
+            final Class protectionDomainBase) {
+        Class clazz = loadClassIgnoringExceptions(loader, actualClassName);
+        if (clazz != null) {
+            return clazz;
         }
 
-        Class result = null;
-        try {
-            if(System.getSecurityManager() == null) {
-                result = _generate(loader, protectionDomainBase.getProtectionDomain(),
-                                   props);
-            } else {
-                result = (Class)  java.security.AccessController.doPrivileged
-                        (new java.security.PrivilegedAction() {
-                    public java.lang.Object run() {
-                        return  _generate(loader, protectionDomainBase.getProtectionDomain(),
-                                          props);
-                    }});
+        synchronized (EJBUtils.class) {
+            clazz = loadClassIgnoringExceptions(loader, actualClassName);
+            if (clazz != null) {
+                return clazz;
             }
-        } catch (RuntimeException runEx) {
-            //We would have got this exception if there were two (or more)
-            //  concurrent threads that attempted to define the same class
-            //  Lets try to load the class and if we are able to load it
-            //  then we can ignore the exception. Else throw the original exception
-            try {
-                result = loader.loadClass(actualClassName);
-                _logger.log(Level.FINE, "[EJBUtils] Got exception ex: " + runEx
-                        + " but loaded class: " + result.getName());
-            } catch (ClassNotFoundException cnfEx) {
-                throw runEx;
-            }
-        }
 
-        return result;
+            cgf.evaluate();
+
+            final Properties props = new Properties();
+            if (_logger.isLoggable(Level.FINE)) {
+
+                props.put(DUMP_AFTER_SETUP_VISITOR, "true");
+                props.put(TRACE_BYTE_CODE_GENERATION, "true");
+                props.put(USE_ASM_VERIFIER, "true");
+
+                try {
+
+                    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                    PrintStream ps = new PrintStream(baos);
+
+                    _sourceCode(ps, props);
+                    _logger.fine(baos.toString());
+
+                } catch (Exception e) {
+                    _logger.log(Level.FINE, "exception generating src", e);
+                }
+
+            }
+
+            Class result = null;
+            try {
+                if (System.getSecurityManager() == null) {
+                    result = _generate(loader, protectionDomainBase.getProtectionDomain(),
+                            props);
+                } else {
+                    result = (Class) java.security.AccessController.doPrivileged(new java.security.PrivilegedAction() {
+                        public java.lang.Object run() {
+                            return _generate(loader, protectionDomainBase.getProtectionDomain(),
+                                    props);
+                        }
+                    });
+                }
+            } catch (RuntimeException runEx) {
+                // We would have got this exception if there were two (or more)
+                // concurrent threads that attempted to define the same class
+                // Lets try to load the class and if we are able to load it
+                // then we can ignore the exception. Else throw the original exception
+                try {
+                    result = loader.loadClass(actualClassName);
+                    _logger.log(Level.FINE, "[EJBUtils] Got exception ex: " + runEx
+                            + " but loaded class: " + result.getName());
+                } catch (ClassNotFoundException cnfEx) {
+                    throw runEx;
+                }
+            }
+
+            return result;
+        }
     }
 
+    private static Class loadClassIgnoringExceptions(ClassLoader classLoader, String className) {
+        try {
+            return classLoader.loadClass(className);
+        } catch (Exception e) {
+            return null;
+        }
+    }
 
     public static RemoteBusinessWrapperBase createRemoteBusinessObject
         (String businessInterface, java.rmi.Remote delegate)

